docs/architecture/bignum.md

```diff
@@ -0,0 +1,129 @@
+                            size_t n, mbedtls_mpi_mod_modulus *N )
+
+int mbedtls_mpi_mod_mul( mbedtls_mpi_mod_residue *X, mbedtls_mpi_mod_residue *A,
+                         mbedtls_mpi_mod_residue *B, mbedtls_mpi_mod_modulus *N )
+
+int mbedtls_mpi_mod_add( mbedtls_mpi_mod_residue *X, mbedtls_mpi_mod_residue *A,
+                         mbedtls_mpi_mod_residue *B, mbedtls_mpi_mod_modulus *N )
+
+int mbedtls_mpi_mod_neg( mbedtls_mpi_mod_residue *X, mbedtls_mpi_mod_residue *A,
+                         mbedtls_mpi_mod_modulus *N )
+
+int mbedtls_mpi_mod_sub( mbedtls_mpi_mod_residue *X, mbedtls_mpi_mod_residue *A,
+                         mbedtls_mpi_mod_residue *B, mbedtls_mpi_mod_modulus *N )
+
+int mbedtls_mpi_mod_prime_inv( mbedtls_mpi_mod_residue *X,
+                               mbedtls_mpi_mod_residue *A,
+                               mbedtls_mpi_mod_modulus *P )
+```
+
+As mentioned above, there will be a second set of arithmetic functions with the same semantics. This second set of functions will be only used in the ECP module and won't perform any sanity checks on the parameters.
+
+There are several occasions where currently other operations are used, but these can be implemented with the above functions. (For example `mbedtls_mpi_sub_int_mod()` is only ever called on 3 and can be replaced by three calls to `mbedtls_mpi_mod_sub()`.)
+
+### Utility Functions
+
+There are a number of utility functions used throughout the ECC modules:
+
+```C
+mbedtls_mpi_get_bit()
+mbedtls_mpi_set_bit()
+mbedtls_mpi_copy()
+mbedtls_mpi_lset()
+mbedtls_mpi_cmp_int()
+mbedtls_mpi_cmp_mpi()
+mbedtls_mpi_safe_cond_assign()
+mbedtls_mpi_safe_cond_swap()
+mbedtls_mpi_random()
+mbedtls_mpi_fill_random()
+mbedtls_mpi_size()
+mbedtls_mpi_bitlen()
+```
+
+For these we need to provide equivalent functionality for the new type while respecting the pre and post conditions described above.
+
+Some of these might not be called directly in the end (for example `mbedtls_mpi_size()` is either used in I/O or can be inferred from the context.)
+
+### Input/Output
+
+We need to provide analogues to the standard input/output functions:
+
+```C
+mbedtls_mpi_write_binary_le()
+mbedtls_mpi_read_binary_le()
+mbedtls_mpi_write_binary()
+mbedtls_mpi_read_binary()
+```
+
+(`mbedtls_mpi_read_string()` is only ever used in selftest and in a function that is not called from anywhere and therefore can be ignored.)
+
+`derive_mpi` in ECDSA does some form of input as well and will need special considerations when switching to the new type.
+
+## Conversion
+
+Beyond the operations above we will need to convert to and from `mbedtls_mpi` as this is what public facing interfaces use. Since `mbedtls_mpi` and `mbedtls_mpi_mod_residue` use the same limb type, conversion might do copy or aliasing. To keep heap consumption low, we will be providing both and using the most suitable method based on the parameter type.
```

Comment:
We also need means to create internal MPIs from raw buffers, e.g. if we want temporaries. See e.g. the ECDSA prototype https://github.com/hanno-arm/mbedtls/blob/ecp_prototype/library/ecdsa.c#L309.

Reply:
I agree, I think that would be part of the normal life cycle of an internal MPI ([see bellow](https://github.com/Mbed-TLS/mbedtls/pull/5895#discussion_r894255809))